packages/scripts/src/upgrade-tests/register-new-contract/2_Verify.ts

```diff
@@ -0,0 +1,173 @@
+  const finder = await getContractInstance<Finder>("Finder");
+  const governor = await getContractInstance<Governor>("Governor");
+  const registry = await getContractInstance<Registry>("Registry");
+
+  const startLookupBlock = (await await registry.provider.getBlockNumber()) - 250; // ~ 1hour ago
+
+  const decodedSubTransactions = decodedData.params.transactions.map((transaction: ProposedTransaction) => ({
+    to: transaction.to,
+    transaction: decodeData(transaction.data),
+  }));
+
+  const governorRootRelays: RelayTransaction[] = [];
+  const governorHubRelays: RelayTransaction[] = [];
+
+  decodedSubTransactions.forEach((relayTransaction: RelayTransaction) => {
+    if (relayTransaction.transaction.name === "relayGovernance") {
+      if (relayTransaction.transaction.params.calls) {
+        governorHubRelays.push(relayTransaction);
+      } else {
+        governorRootRelays.push(relayTransaction);
+      }
+    }
+  });
+
+  const registryL1Calls = decodedData.params.transactions.filter(
+    (transaction: ProposedTransaction) => transaction.to === registry.address
+  );
+
+  const registerContractTransactions = registryL1Calls.find(
+    (transaction: ProposedTransaction) => decodeData(transaction.data).name === "registerContract"
+  );
+
+  const registerTx = decodeData(registerContractTransactions.data);
+
+  const newContractAddressMainnet = registerTx.params.contractAddress;
+
+  const newContractAddressCheck = await getAddress(newContractName, Number(networkId));
+
+  console.log("Verifying that new contract address is correct...");
+  assert.equal(newContractAddressMainnet, newContractAddressCheck);
+
+  const governorRootTunnel = await getContractInstance<GovernorRootTunnel>("GovernorRootTunnel"); // for polygon
+  const governorHub = await getContractInstance<GovernorHub>("GovernorHub"); // rest of l2
+
+  console.log("Verifying GovernorHub relays...");
+  for (const relay of governorHubRelays) {
+    await verifyGovernanceHubMessage(governorHub, relay, startLookupBlock);
+  }
+  console.log("Verified!");
+
+  console.log("Verifying GovernorRootTunnel relays...");
+  for (const relay of governorRootRelays) {
+    await verifyGovernanceRootTunnelMessage(governorRootTunnel, relay, startLookupBlock);
+  }
+  console.log("Verified!");
+
+  console.log("Verifying that Governor doesn't hold the creator role...");
+  !(await registry.holdsRole(RegistryRolesEnum.CONTRACT_CREATOR, governor.address));
+  console.log("Verified!");
+
+  console.log("Verifying that the New Contract is registered with the Registry...");
+  assert(await registry.isContractRegistered(newContractAddressMainnet));
+  console.log("Verified!");
+
+  console.log("Verifying that the New Contract is registered with the Finder...");
+  assert.equal(
+    (await finder.getImplementationAddress(hre.ethers.utils.formatBytes32String(newContractName))).toLowerCase(),
+    newContractAddressMainnet.toLowerCase()
+  );
+  console.log("Verified!");
```

Comment:
so we dont actually verify that on the L2 things are done correctly, right? would we be able to run this against a hardhat fork of say Polygon to check that polygon has correctly added the contracts, via the cross-chain actions? i.e that the cross-chain addition worked correctly by re-running this on each chain? we might need to add case work for this (or a separate script). wdyt?

Reply:
Yes, I think you are right, I will add an extra script to this!